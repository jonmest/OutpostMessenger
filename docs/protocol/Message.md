# Message protocol
For sake of simplicity, we will define a protocol for the structure of plaintext and ciphertext messages exchanged between clients.

## Plaintext message
```
{
    'recipient' : String,
    'sender'    : String,

    'message'   : {
        'count' : integer, // Start count at 1, always increase
        'plain' : String,
        'date'  : String
    }
}
```

## Ciphertext message
```
{
    'recipientId' : String,
    'nonce': String,
    'secret' : {
        'sender'    : String,
        'date'  : String,
        'plain'   : Asymmetric nonce + encrypt(JSON.Stringify(message))
    } => String

}
```