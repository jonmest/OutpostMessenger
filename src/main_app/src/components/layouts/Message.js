import React, { Fragment } from 'react'
import Moment from 'react-moment'
import ReactMarkdown from 'react-markdown'

const Message = ({ data, sender, contact, timestamp, me }) => {
    const time = new Date(timestamp)
  
    return (
        <Fragment>                   
            <div className="columns">
            <div className="column">
            {
                (sender === contact.id ) ?
                <Fragment>
                <span className="help has-text-grey-light">
                <Moment fromNow>{ time }</Moment>
                </span>
                <article className="message">
                <div className="message-body">
                    <ReactMarkdown source={ data }/>
                </div>
                </article>
                </Fragment> :
                null
            }
            </div>
            <div className="column">
            {
                (sender !== contact.id || me === contact.id ) ?
                <Fragment>
                    <span className="help has-text-grey-light">
                        <Moment fromNow>{ time }</Moment>
                    </span>
                    <article className="message  is-dark">
                        <div className="message-body">
                            <ReactMarkdown source={ data }/>
                        </div>

                    </article>

                </Fragment> :
                null
            }
            </div>
            </div>
        </Fragment> 
        )
}


export default Message