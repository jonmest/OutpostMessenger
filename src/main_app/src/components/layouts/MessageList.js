import React, { Fragment } from 'react'
import Message from './Message'

const MessageList = ({ messages, contact, me }) => {
    return (
        <Fragment>                   
            {
          messages.filter(item => item !== undefined).map((item, index) => {
            return <Message key={index}
            data={item.data}
            sender={item.sender}
            contact={contact}
            timestamp={item.timestamp}
            me={me}
            />
          })
          }
        </Fragment> 
        )
}


export default MessageList