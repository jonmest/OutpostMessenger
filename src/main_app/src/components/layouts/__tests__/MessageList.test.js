import React from 'react';
import Message from '../Message'
import MessageList from '../MessageList'
import renderer from 'react-test-renderer';

const date = new Date()
const now = `${date.getFullYear()}-${date.getMonth() + 1}-${date.getDate()} ${date.getHours()}:${date.getMinutes()}:${date.getSeconds()}.000`
test('Matches snapshot', () => {
    const contact = "Thomas"
    const me = "Timothy"
    const messages = [
        {
            data: "This is a message.",
            sender: contact,
            contact: contact,
            timestamp: now,
            me
        },

        {
            data: "This is another message.",
            sender: me,
            contact: contact,
            timestamp: now,
            me
        }
    ]
    let component = renderer.create(
        <MessageList messages={messages} contact={contact} me={me}/>,)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
})
