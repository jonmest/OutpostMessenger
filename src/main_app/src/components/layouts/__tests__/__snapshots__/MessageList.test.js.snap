// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
Array [
  <div
    className="columns"
  >
    <div
      className="column"
    />
    <div
      className="column"
    >
      <span
        className="help has-text-grey-light"
      >
        <time
          dateTime={"2020-05-12T08:23:25.000Z"}
        >
          a few seconds ago
        </time>
      </span>
      <article
        className="message "
      >
        <div
          className="message-body"
        >
          <p>
            This is a message.
          </p>
        </div>
      </article>
    </div>
  </div>,
  <div
    className="columns"
  >
    <div
      className="column"
    />
    <div
      className="column"
    >
      <span
        className="help has-text-grey-light"
      >
        <time
          dateTime={"2020-05-12T08:23:25.000Z"}
        >
          a few seconds ago
        </time>
      </span>
      <article
        className="message "
      >
        <div
          className="message-body"
        >
          <p>
            This is another message.
          </p>
        </div>
      </article>
    </div>
  </div>,
]
`;
